### ISSUE TITLE
Proposal to separate media focus from HTMLMediaElement objects

### ISSUE THREAD

**Issue by richtr** [richtr/html-media-focus#5]
@marcos wrote:

> > Just thinking out loud, but something like where both `RemoteControl` and `HTMLMediaElement` implement [NoInterfaceObject] `MediaKeyReceiver`.
> > 
> > This would allow MediaKeyReceiver to remain as an independent API, but HTMLMediaElement could still be explained (as far as media-keys are concerned) in terms of `MediaKeyReceiver`. A constructable, or requested, `RemoteControl` will hopefully then allow the ability to address use cases like controlling RevealJS presentations, the Flash use cases, and things built with custom elements, etc.
> > 
> > `RemoteControl` would have to be requested (`requestRemoteControl()`), granting the user the explicit ability to now allow arbitrary web apps to claim the media focus if they are not actually controlling media (or if they are playing media through a third-party plugin, as per Flash). Hopefully requiring an explicit grant should also prevent media-key hijacking.
> 
> Would something like that work?

**Comment by richtr** [comment]
Some feedback on this:
- If `HTMLMediaElement` implements `MediaKeyReceiver` would that
  _enforce_ playback behavior on such media objects? It feels like the
  proposal is to duplicate existing HTMLMediaElement events just for the
  sake of having both `RemoteControls` and `HTMLMediaElement` implement
  from the same interface. I'm not sold on the value of that. The
  HTMLMediaElement machinery is already ready to work with remote
  control events and we could just reuse that when possible.
- Would events be fired only at _focused_ HTMLMediaElement objects?
  Does an HTMLMediaElement object need to hint that it is
  _focusable_ or does this apply equally to all HTMLMediaElement objects?
- Permission dialogs. Not a fan :)
- With a standalone API the page will need to also know when it loses
  focus. Both that and some hints to know when would be a good time to
  re-request focus (when another web app has finished with it...so a web
  app may also need a way to release media focus also) would be good. It
  feels a bit icky that a page could arbitrarily interrupt another web
  app via this API. Would that logic be obvious to an end user? Will web
  apps compete in an endless loop for media focus between themselves?
  Will greedy web pages try to sit on media control focus indefinitely (taking 
  away controls from e.g. native applications too)?
- `MediaKey`... is already used extensively in the [EME API](https://w3c.github.io/encrypted-media/) so
  we should probably come up with a different interface name and avoid
  using this term wherever we can.

I'm not convinced based on the above that the lifecycle of media focus is all that well
thought out but then I suspect we need some prototyping to confirm or
deny that.

**Comment by domenic** [comment]
This seems pretty reasonable although I'm unsure why there'd be a MediaKeyReceiver mixin instead of HTMLMediaElement just having a `.remoteControl` property (or perhaps a privately-held instance of RemoteControl that it doesn't even give access to).

**Comment by marcoscaceres** [comment]
@domenic, I guess it could be explained through a privately held instance. 

The thing that `HTMLMediaElement` is missing is to somehow handle the skip-forward and skip-back action in the same manner it handles play/pause (which would require `HTMLMediaElement` to support a playlists - i.e., a list of sources). If we solve that, then `RemoteControl` can be privately-held in `HTMLMediaElement` without needing to expose it at all (as it would all "just work"). 

@richt, how does your proposal handle forward and back commands coming in from the OS?

**Comment by foolip** [comment]
Excellent, I think tying this less tightly to `HTMLMediaElement` is a good thing. However, I would take it a step further and, like @domenic suggests, simply make it a separate object, kind of in the style of `MediaController`.

CC @jernoble.

**Comment by marcoscaceres** [comment]
Ok, so, how about: 

If the `RemoteControl` is internalized in `HTMLMediaElement`, then `RemoteControl`. A is declared as remote controllable as `<video remotecontrolled>`:
1. When the `remotecontrolled` attribute is set, an internal instance of `RemoteControl` is constructed. Listener for events are registered for remote control event ("play", "pause", "forward", "back"). 
2. user presses play button through a `<video>` element's  shadow DOM - which causes .play() to start.
3. user presses play\pause through on headset. 
4. `RemoteControl` receives the event "pause" as response of headset button being pressed (we know it's a "pause" because we know when media is playing through the UA):  
   1. If the `HTMLMediaElement.paused` is `true`, then `.play()` it . Otherwise, `.pause()` it.
   2. Normal HTML steps apply from here: a task is queued to fire a simple event on `video` named `pause`. 
5. Forward and rewind do nothing for now... till they are supported in the future. 

If we now request an additional RemoteControl() through script, it receives the remote control events too: 

``` JS
navigator.requestRemoteControl().then(
   rc => rc.addEventListener("mediakeys", controlHandler)
); 
function controlHandler(e){
   switch(e.type){
       case "forward": 
        $("#video").src = nextVideo(); 
        // and so on... 
   }
}
```

**Comment by foolip** [comment]
Yeah, this is beginning to sound more like what I had in mind. However, when is the promise returned by `requestRemoteControl()` resolved?

Inspired by @jernoble in https://github.com/whatwg/media-keys/issues/1#issuecomment-73090289 and some code that WebKit has gained after the Blink fork dealing with precisely these concepts, I'm tempted to call it `MediaSession`, since it's not merely about remote controls and would do something useful even on hardware without the relevant buttons.

A `MediaSession` would have a set of `HTMLMediaElement`s or `AudioContext`s, and it is when one of the members in this set reaches a particular state (approximately when the `playing` event is fired in the case of media elements) that this `MediaSession` becomes the recipient of hardware keys. It would continue to receive these events as long as nothing else takes the "focus" away, but would also have a `.end()` or similar to explicitly release control over lock screen UI and similar.

By associating multiple audio-producing objects with a single `MediaSession`, we avoid the problem of handing "audio focus" between media elements in the multi-element playlist case.

**Comment by foolip** [comment]
Also, I would like for there to be a document-wide default `MediaSession` to which all media elements belong by default, and which gets its title from document.title or similar, explaining the automatic lock screen behavior iOS already has and Opera for Android's notification UI for ongoing media playback.

**Comment by foolip** [comment]
So, something with this approximate shape:

``` webidl
enum MediaSessionState { "paused", "playing", /* maybe more */ };
typedef (HTMLMediaElement or AudioContext) MediaSessionMember;

interface MediaSession : EventTarget {
    readonly attribute MediaSessionState state;

    readonly attribute MediaSessionMember[] members;
    void addMember(MediaSessionMember member);
    void removeMember(MediaSessionMember member);

    void end(); // or inactivate() maybe

    // events to notify
    attribute EventHandler onactive; // session became the active one
    attribute EventHandler oninactive; // session is no longer active

    // events with default actions which can possible by preventDefault()'ed
    attribute EventHandler onplay; // forwards to members by default
    attribute EventHandler onpause; // forwards to members by default
    attribute EventHandler onprevious; // previous track
    attribute EventHandler onnext; // next track
    attribute EventHandler onstar; // maybe stuff like this that iOS has
};
```

**Comment by jernoble** [comment]
@richtr 

> With a standalone API the page will need to also know when it loses
> focus. Both that and some hints to know when would be a good time to
> re-request focus (when another web app has finished with it...so a web
> app may also need a way to release media focus also) would be good.

Just for a point of reference, a similar situation exists on iOS. An audio session can be interrupted at any time. E.g., Spotify can be interrupted by the Music app, or Spotify can be interrupted by a phone call. In one case, you definitely don't want Spotify to resume when the Music app ends its interruption. ("Pausing my music plays Spotify!?")  But you do want Spotify to resume when the phone call ends.

The way this is dealt with on iOS is, when an interruption ends, the notification to the interrupted app contains a `shouldResume` boolean. The system generally has more information available about what types of audio playback caused and received the interruption, and can set the rules about how these overlapping audio sessions interact.

As far as WebKit on iOS is concerned, we simply obey the `shouldResume` request when a HTMLMediaElement (or, as @foolip notes, a MediaSession) is notified that interruption ends.

**Comment by marcoscaceres** [comment]
> Yeah, this is beginning to sound more like what I had in mind. However, when is the promise returned by requestRemoteControl() resolved?

When user grants it access. This is why it would be nice to be able to get the `.remoteControl` from `HTMLMediaElement`. Then, if you want implement a playlist, you don't need to request it from the user. 

However, if you want to build something like RevealJS, then you can request remote control access from the user. 

> I'm tempted to call it MediaSession,

I don't have any strong opinion about the name. We can bikeshed like crazy later. Let's just get the model right. 

> A MediaSession would have a set of HTMLMediaElements or AudioContexts, and it is when one of the members in this set reaches a particular state (approximately when the playing event is fired in the case of media elements) that this MediaSession becomes the recipient of hardware keys. 

Mozilla is pushing for having these events routed to an app without `HTMLMediaElements` or `AudioContexts`, hence the `requestRemoteControl()` thing. This is to then pass the events from the web page to Flash through Flash's `sendMessage()` API or whatever.  

> ```
>     attribute EventHandler onstar; // maybe stuff like this that iOS has
> ```

Let's not go here yet - minimum set for interop and to address the use cases first.

**Comment by jernoble** [comment]
@marcosaceres

> When user grants it access

Explicit user permission is the security model of last resort. We should try to avoid it if at all possible. 

EDIT:

> However, if you want to build something like RevealJS, then you can request remote control access from the user.

If we want to enable general purpose remote control events, we should do that in another spec, and not shoe-horn it into the media-focus (or media-keys) spec.

**Comment by foolip** [comment]
I agree with @jernoble, basically. Getting access to media keys without media playback doesn't even seem possible on iOS and even if that's the only "problem" it seems like a pretty sane model, and adhering to it doesn't seem to get in the way of any of the use cases I started out with.

All else equal I think we could support Flash, but AFAICT it would actually make the API worse in the other cases. A bolt-on solution could be adding `HTMLEmbedElement` and `HTMLObjectElement` to `MediaSessionMember` and defining the conditions under which those objects can cause a `MediaSession` to become active, possibly by requiring it to send something to an audio output device or showing a permissions dialog. Implementors who don't support Flash could then simply ignore these bits of the API.

**Comment by foolip** [comment]
@marcoscaceres 

> it would be nice to be able to get the .remoteControl from HTMLMediaElement

Yep, that seems nice and easy with any solution that has an object separate from `HTMLMediaElement` as the "focus holder". In the `MediaSession` case, we would have `HTMLMediaElement.session`, rather a lot like `MediaController` and `HTMLMediaElement.controller`.

**Comment by marcoscaceres** [comment]
@jernoble, based on the feedback/discussion so far, I've asked Mozilla to reconsider their position on that the [Flash use case is a "MUST"](https://bugzilla.mozilla.org/show_bug.cgi?id=1112032#c24). 

Please see:
https://bugzilla.mozilla.org/show_bug.cgi?id=1112032#c33

I expect @sicking might have some strong counter arguments.

**Comment by richtr** [comment]
> @richt, how does your proposal handle forward and back commands coming in from the OS?

There's a section in the [README](https://github.com/richtr/html-media-focus/blob/gh-pages/README.md#what-about-the-forward-and-backward-media-control-interface-keys) that covers this. Essentially we proposed to fire appropriate events at the _focused_ `HTMLMediaElement` object and the web page can choose to respond to those events or not. The behavior is left up to the web page. Those buttons could be wired to either skip forward in the current track by e.g. 15 seconds or to be used to skip to the next track. There is some example event handling code [here](https://github.com/richtr/html-media-focus/blob/gh-pages/index.html#L118-L139).

There is a question of how we keep media focus during a HTMLMediaElement transition (either by changing the .src of the HTMLMediaElement or transferring focus to another HTMLMediaElement object on the web page). I wonder if we could provide bridging media control focus and then after e.g. 5 seconds of inactivity from the page, detach the page from any remote controls.

**Comment by richtr** [comment]
> > However, if you want to build something like RevealJS, then you can request remote control access from the user.
> 
> If we want to enable general purpose remote control events, we should do that in another spec, and not shoe-horn it into the media-focus (or media-keys) spec.

Agreed. 

It is good to discuss general purpose remote control events but it does introduce a whole host of new issues, i.e.:
- when to request focus/session, 
- when to grant focus/session, 
- when to release focus/session, 
- when to re-request focus/session, 
- when to re-grant focus/session,
- whether to resume playback or keep media paused when media focus/session is returned, 
- how the platform could enforce/reflect the state of `AudioContext`s and Flash-based media, 
- how to design and implement permission dialogs for this.

That is actually _a lot_ of variables.

The alternative of hooking directly on to `HTMLMediaElement` and/or implementing `MediaSession` that relies on attached `HTMLMediaElement` objects removes all of these issues as they can be bound with the HTMLMediaElement machinery.

That is not to say one solution is necessarily better than the other but we can say general remote control events is orders of magnitude more complex than the other. 

We might get further quicker if we look at this as two (or more) specifications.
